### DISCUSSION TITLE
Discussion: Inferred "Quantum" Lambdas

### DISCUSSION THREAD

**Discussion by HaloFour** [dotnet/csharplang#1694] [Language Ideas] [answered]
Lambda expressions are currently not supported for local variable inference.  This is because the signature of the lambda expression does not indicate the type of the delegate which is necessary to create an instance of that delegate type.

What if the compiler didn't have to actually know the delegate type at the time that the variable is declared?  What if the variable could have a form of "pseudo-type" based on its signature and the compiler could "collapse the wave function", as it were (hence the name), at the point that the instance needed to be converted to a delegate?

Here's an example with captures, which isn't very different:

```cs
public void M(bool b, int x) {
    var f = (int y) => x == y;
    if (b) {
        M1(f);
    }
    else {
        M2(f);
    }
}

public void M1(Func<int, bool> f) { }
public void M2(Predicate<int> p) { }

// under the hood

[CompilerGenerated]
private class $DisplayClass {
    public int x;

    public bool $M(int y) {
         return x == y;
    }
}

public void M(bool b, int x) {
    var displayClass = new $DisplayClass();
    displayClass.x = x;
    if (b) {
        Func<int, bool> $temp1 = displayClass.$M;
        M1($temp1);
    }
    else {
        Predicate<int> $temp2 = displayClass.$M;
        M2($temp2);
    }
}
```

Here's an example without captures:

```cs
public void M(bool b) {
    var f = (int y) => x % 2 == 0;
    if (b) {
        M1(f);
    }
    else {
        M2(f);
    }
}

public void M1(Func<int, bool> f) { }
public void M2(Predicate<int> p) { }

// under the hood

[Serializable]
[CompilerGenerated]
private sealed class $C {
    public static readonly $C _c = new $C();

    public static Predicate<int> _p;
    public static Func<int, bool> _f;

    internal bool $M(int x) {
        return x % 2 == 0;
    }
}

public void M(bool b) {
    if (b) {
        Predicate<int> $temp1 = $C._p ?? ($C._p = $C._c.$M);
        M1($temp1);
    }
    else {
        Func<int, bool> $temp2 = $C._f ?? ($C._f = $C._c.$M);
        M2($temp2);
    }
}
```

Advantages:
1.  Inferred local variables for lambda expressions
1.  Marginally less verbose
1.  Avoiding the double-delegate invocation if trying to pass to multiple functions accepting different but compatible delegates

Disadvantages:
1.  An additional allocation for every delegate type that the lambda is "collapsed" to, although this is already unavoidable
1.  The variable is some weird pseudo-type which makes it unusable in other contexts, e.g. `f.GetType()`.
1.  This is similar to using local functions as the delegate target

While typing this up I believe I've come to the conclusion that it's not a terribly good idea.  But whatever, tear it up folks!

**Comment by bbarry** [comment]
I like your unspeakable types notation. $DisplayClass is easier to type than something like `<>t__M__DisplayClass` and almost doesn't break github markdown syntax highlighting.

**Comment by HaloFour** [comment]
@bbarry   Not to mention you can read it without trying to pick the dots out of a wall of angle brackets and underscores.

**Comment by agocke** [comment]
I believe C++ does something similar with `auto` and both lambdas and array literals and oh-my-god is it frustrating. I think there's another proposal for assuming a default type (Func/Action or maybe we could introduce a "real" function type into the language) which IMHO is a much better experience.

Trying to get a literal to target-type to the appropriate thing is hard enough when the literal is actually inline, but when it's in a variable you end up with "spooky action at a distance" (to continue the quantum superposition metaphor 😉).

**Comment by HaloFour** [comment]
@agocke 

> I think there's another proposal for assuming a default type

I largely agree.  The only problem with those assumptions is that `Func<...>`/`Action<...>` only work for the simplest kinds of delegates.  Toss in a `ref` parameter and the compiler could no longer infer the delegate type.  You also have the issue where if you assign the lambda to an inferred local and then pass that to a method expecting some other kind of delegate that you end up with another allocation and additional overhead on invocation.  Not that I think that either of these problems would be show-stoppers.

**Comment by agocke** [comment]
I think the cleanest solution would be inferring a native function type: `var f = x => x;` infers a type of `T -> T`. Then there would be conversions to delegate types. The problem, of course, is in the details. I'm not sure there's any off-the-shelf inference algorithm that will be compatible with C#'s overload resolution rules. If someone wants to take a stab at adapting Damas-Milner for C# conversion rules, that would be great.

**Comment by bondsbw** [comment]
Internally there would need to be a more generalized concept of lambdas than exists today (`int -> bool`) which is convertible to any compatible delegate type.

In that case, why not expose it somehow?

```c#
lambda(int => bool) f = (int y) => x == y;
```

Then (perhaps) you could return it, pass it as a parameter, declare it as a field, etc.

**Comment by agocke** [comment]
If you have to write the type, the point is lost anyway. Someone still needs to work out the inference algorithm.

**Comment by bondsbw** [comment]
That could be true for any `var`.  Granted this one is definitely more complex.

(Edit... I say it is more complex, but I'm not sure I really believe it.  The inferred type of `var` would always be a `lambda(...)`.  The target-typed conversion would be the complex part, but is it really that hard?)

**Comment by bondsbw** [comment]
If we really don't want a new keyword:

```c#
delegate(int => bool) f = (int y) => x == y;
```

**Comment by bbarry** [comment]
@bondsbw what about `Expression<Func<int, bool>>`

**Comment by bondsbw** [comment]
@bbarry Like `Func<int, bool>` you wouldn't have the ability to declare `ref` parameters.

(assuming you meant `Expression<Func<int, bool>>`)

**Comment by bbarry** [comment]
yes (edited)

I mean `delegate(int => bool) f` seems to imply that the `f` variable is denoting a compiled lambda expression. What I suspect people who want `var f = y => x == y;` desire is something like "declare `f` as a lambda matching this thing however it winds up getting used..."

I'd propose to do exactly that:

the `declaration_statement` consisting of

    'var' identifier '=' lambda_expression ';'

could simply exist and be inlined by the compiler wherever that identifer exists in scope (edit: as @agocke implies there is a devil here in identifying type information within this statement).

(not actually suggesting this, I personally think this proposal is of very little value)

**Comment by bondsbw** [comment]
> What I suspect people who want `var f = y => x == y;` desire is something like "declare `f` as a lambda matching this thing however it winds up getting used..."

That's precisely what `delegate(int => bool) f` does.

**Comment by bondsbw** [comment]
An `Expression` could be rendered downstream once it is converted.

**Comment by bbarry** [comment]
Why should there be any additional syntax then for that?

Why not merely relax the language from:

> * The initializer *expression* must have a compile-time type.

to:

> * The initializer *expression* must have a compile-time type or be a *lambda_expression*.

and then add something like:

If the initializer *expression* is a *lambda_expression*, the following process is invoked:

1. Let the *identifier* be `v` 
2. where the variable `v` is used in scope, determine type of variable. It is an error if the type cannot be determined.
3. for each unique usage type, replace the variable `v` with a generated id and duplicate the *declaration_statement* with the new type and reparse.

**Comment by bondsbw** [comment]
@bbarry That seems reasonable for the main proposal.

> Why should there be any additional syntax then for that?

Implicit types (`var`) aren't allowed as parameter types, return types, or members.

**Comment by erik-kallen** [comment]
Could we write `var f = y => x == y;` and have the type inferred when the wave function is collapsed?

**Comment by Neme12** [comment]
I would prefer just having something like `var f = (Foo x) => x.ToString()` and having first class function types. Being explicit about types in a lambda might not be as nice but this is still C# and the types have to be specified somewhere. I find this nicer than `Func<Foo, string> f = x => x.ToString()`

Maybe I don't quite understand the proposal but I don't see how this would work. Would reassignment be prohibited? If that's the case, combined the fact that the delegate is created at the point of use, this isn't really much different from a local function (with added limitations).

**Comment by Neme12** [comment]
> If you have to write the type, the point is lost anyway. 

What point is lost? I don't think having to write types in the argument list would be such a nuisance. I find that much more preferrable than `Func<>`.

> The problem, of course, is in the details. I'm not sure there's any off-the-shelf inference algorithm that will be compatible with C#'s overload resolution rules. If someone wants to take a stab at adapting Damas-Milner for C# conversion rules, that would be great.

For better or worse, this is what we have with C# and as much as we can sometimes appreciate this sort of type inference in other languages, I wouldn't want to introduce that in just one specific corner of C#.

**Comment by bondsbw** [comment]
> Would reassignment be prohibited?

It's `var` so I would say no.  Prohibiting reassignment should use a different keyword.

**Comment by Neme12** [comment]
> Prohibiting reassignment should use a different keyword.

I agree. If we have readonly locals, I'd like an actual feature with its own syntax rather then saying "most variables are mutable, unless you have this inferred lambda thing (which doesn't seem to fit into the language)".

**Comment by CyrusNajmabadi** [comment]
Here's my take:  I don't like the idea of the 'collapsed wave function'.  But i do like the idea of a shorthand way of writing functions/actions.  So i would propose the following:

If the user writes a lambda, and there is no contextual delegate type, then we attempt to create a Func/Action for that lambda based on the type information we do have at declaration time.  Examples of that woul dbe something like:

```c#
var f = (Foo x) => x.ToString()
```

Now, what do you if you really wanted to write:

```c#
var identity = x => x;
```

You don't have types to give there (because this is fundamentally something generic).  So i would propose allowing you to do what TypeScript lets you do.  Just write this as:

```c#
var identity = <T>(T x) => x;
```

In other words, allow one to declare a lambda not just with parameter types, but with generic type parameters as well.  

The algorithm here would be: take the lambda and use the type-parameters+parameters to determine the delegate signature.  Analyze the body with those parameters to determine what the return value is (if any).  If you have a return value, make an appropriate ```Func<...>``` with those types.  If there is nothing returned, then make it an ```Action<...>```.

**Comment by CyrusNajmabadi** [comment]
In other words, there is no 'inferred quantum lambda'.  We just imbue the language with knowledge of Func/Action, and we make it pretty simple to provide shorthand literals for those most common of delegate types.  

Note: a downside of this is that you wouldn't be able to do something like:

```c#
var x = (ref int x) => ...;
```

But i honestly don't care.  I think there is sufficient value in just being able to write most of your Func/Action lambdas without having to specify those types, versus being able to write a lambda for any and all delegate types.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

So the following would be a compiler error?

```cs
void M(Predicate<string> p) { }

var predicate = (string s) => s != null;
M(predicate);
```

**Comment by CyrusNajmabadi** [comment]
Yes.  IMO, as the majority of APis are now written with Func/Actions, this will be less impactful.  My suggestion on most APIs that are written against older delegate types is that they provide a Func/ACtion overload (or that such a thing is provided with an extension).

**Comment by Neme12** [comment]
> Yes. IMO, as the majority of APis are now written with Func/Actions

I wish. I just had to use `Predicate`s today for a UI framework. These sorts of things stick around (for compatibility?) even in newer code bases. And some people are unaware that the guidance is to use `Func` and/or find `Predicate` more readable.

**Comment by Neme12** [comment]
> My suggestion on most APIs that are written against older delegate types is that they provide a Func/ACtion overload

Wouldn't that be an ambiguity for the caller unless they cast to the appropriate type?

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

I disagree with the notion that delegate types are somehow "older" or "legacy".  I continue to use them for all but the simplest cases as being nominal with proper parameters makes them self-documenting, plus I can slap XML docs on them. I also frequently use them for injected dependencies where an interface/implementation would be overkill.

**Comment by Neme12** [comment]
> I disagree with the notion that delegate types are somehow "older"

Well, they are older... that's just a fact of reality :smile:

**Comment by HaloFour** [comment]
@Neme12

> Well, they are older... that's just a fact of reality 😄

Many of my delegates are quite a bit younger than `Action<...>` or `Func<...>`, thanks.  :P

**Comment by bondsbw** [comment]
@CyrusNajmabadi 
> We just imbue the language with knowledge of Func/Action, and we make it pretty simple to provide shorthand literals for those most common of delegate types.

What does this really buy over imbuing the language with knowledge of the underlying method group?

**Comment by CyrusNajmabadi** [comment]
> What does this really buy over imbuing the language with knowledge of the underlying method group?

it would be much simpler to provide, for one.

**Comment by CyrusNajmabadi** [comment]
> I disagree with the notion that delegate types are somehow "older" or "legacy". I continue to use them for all but the simplest cases as being nominal with proper parameters makes them self-documenting, plus I can slap XML docs on them. I also frequently use them for injected dependencies where an interface/implementation would be overkill.

The question is not if they're used or not.  The question is: is there enough value from just getting Func/Action working as to make this 'good enough'.  IMO, i think it would.  For the other cases, you can still do what you do today.  i.e. using a local function, or just using a named-type when supplying the lambda.


---

Note: i completely agree the quantum-lambda approach is more powerful and would handle more scenarios.  But i also think it's less likely to happen, and the power it provides isn't worth the cost.  So i'm reproposing narrowing things down to something that would be quite easy to provider, and which which would support many scenarios.

Also, note that my suggestion doesn't not eliminate the possibility of doing quantum lambdas.  Indeed, they are complimentary.

**Comment by bondsbw** [comment]
@CyrusNajmabadi 
> Also, note that my suggestion doesn't not eliminate the possibility of doing quantum lambdas. Indeed, they are complimentary.

If `var f` is resolved to `Func<int, bool>` in C# v.X, wouldn't it prohibit changing the definition in v.X+1 to resolve `f` to the method group?

**Comment by Neme12** [comment]
> I continue to use them for all but the simplest cases as being nominal with proper parameters makes them self-documenting, plus I can slap XML docs on them.

I agree. For one thing, Predicate reads better than the word "func". But putting that aside, I can see why it's nice to name the delegate sometimes and especially name the arguments or even provide a doc.

I think in an ideal world, there would be first-class function types, and if you wanted to name them, you would use global type-aliases. But they'd still be compatible. And many times, you wouldn't even need that as much if these function types had optional names similar tu tuples... What a nice dream :smile:

**Comment by CyrusNajmabadi** [comment]
> If var f is resolved to ```Func<int, bool>``` in C# v.X, wouldn't it prohibit changing the definition in v.X+1 to resolve f to the method group?

'var f' woudl be resolved to ```Func<int, bool>``` if you actually wrote:   ```var f = (bool x) => ...```.  So i don't see how method-groups actually come into consideration here.

**Comment by ghost** [comment]
@HaloFour 
> Lambda expressions are currently not supported for local variable inference. This is because the signature of the lambda expression does not indicate the type of the delegate which is necessary to create an instance of that delegate type.

Alternatively, there is possible to extend type infer for lambdas via something like the factory pattern.
```c#
var a = New.Func((int x) => x);
var b = New.Func((int x) => x == 5);
```
Instead
```c#
var a = (Func<int,int>)(x => x);
var b = (Func<int,bool>)(x => x == 5);
```
The implementation is simple (see [a online sample](https://sharplab.io/#v2:CYLg1APgAgDABFAjAbgLACgNQMwIExwDCGA3hnBQrlACxwCyAFAJRzmVnqXdwBuAhgCc4/OAF44AOQCmAdwB0UAKyNGASwB2AFzgAPVmIB8e5mi48KA4QCNxUuYpXrtJ8cd3iJS0+wsB6PysRO0ZlAB5NLQAaSMNmRg8jEzMLCgCg2wlQpQjtKOsAewKAGziEtz1POG8U1N8eIIBjOxkFJDxVSNckpABOBOYfcwsg4BaHds6XfQrE6qHuAF8MZcx0HHx7WVJ6jaQANgQcgC1jcNPssNO4ADMDYxvaij3EQ/CAQSi4a4+vi9/vg97rcnlQEK8jmFPnAAEJ/M45aFwwGXJHw27Ax71AJweR4+q7agQ9rgjok/jA/igl5vRB4KFnOkM0JM97GCkVKmE8G0+lIxl8r4wwwswWw9mU0E4vHyFZAA=))
```c#
public class New
{
    public static Func<Z> Func<Z>(Func<Z> f) => f;
    public static Func<A, Z> Func<A, Z>(Func<A, Z> f) => f;
    public static Func<A, B, Z> Func<A, B, Z>(Func<A, B, Z> f) => f;
    // ...
    
    public static Action Action(Action a) => a;
    public static Action<A> Action<A>(Action<A> a) => a;
    public static Action<A, B> Action<A, B>(Action<A, B> a) => a;
    // ...
}
```
The full implementation is [here](https://bitbucket.org/Makeloft/ace/src/91837895ae7486db17b080161dd6fde7944ff0a0/Ace.Base/Sugar/New.cs?at=master&fileviewer=file-view-default).

Posted by @makeman-from-makeloft

**Comment by bondsbw** [comment]
@Makeloft Check what you quoted.  Your solution forces a delegate type, which this proposal is trying to avoid.

**Comment by ghost** [comment]
@bondsbw Yes, I understand and agree with your remark.

I just show a way how to achieve following advantages on the current language level

> 1. Inferred local variables for lambda expressions
> 2. Marginally less verbose


Posted by @Makeman-from-Makeloft

**Comment by tannergooding** [comment]
The comments from `ghost` have been hidden as `spam`. They are coming from a user who was previously banned and who is currently trying to circumvent their ban.

As a general note: Repeatedly advertising your own repository as a solution to most language proposals, after being warned, is deemed to be "unprofessional conduct"; as is attempting to circumvent a ban.

The Code of Conduct is expected to be followed by all contributors/posters to The `dotnet` organization on GitHub. For more details, please see: https://dotnetfoundation.org/code-of-conduct

**Comment by ghost** [comment]
@tannergooding

> "unprofessional conduct"

1.  Is it professional behavior to ban a user who provide neutral but argumented critics of some LDT decisions?

2. Is it a professional approach to play into politics instead attentive considering of technical proposals?

3. Is it professional to limit free and workable solutions? Delete useful comments, constantly name the author like "spammer", etc...

My circumvent is protest.

> "advertising"

1. It is free open source code.

2. Own repository is a place where I sure that solutions will not be removed. Only the LDM forces me to attach links on the repository to my technical posts regularly because constantly delete them.

Excuse me for this offtop.

Posted by @Makeman-from-Makeloft

**Comment by Thaina** [comment]
I think lambda should be quantum and anonymous all along. It was a mistake to have it must be typed when it could have duplicate signature

We should starting to deprecate the old delegate system

**Comment by Rekkonnect** [comment]
The only problem I see with this is the confusion of `var` in this context. `var` could imply that the target type is inferred, but our purpose is not binding the exact delegate type, probably ever. The only information we want to extract from the lambda expression (or method group) is its arguments and return types, basically the signatures. In this case, new syntax should probably be introduced.

What I would be in favor of is `<keyword> var <identifier> = <expression>`, where `<keyword>` should be a keyword for this exact purpose and the grammar mandates that both the keyword and `var` be used for this scenario.

The only options I could make sense with:
- `delegate var`, which I find a bit too verbose
- `partial var`, with `partial` meaning "partially bound delegate type"
- `free var`, with `free` meaning "freely bound delegate type"

I'm pretty sure there can be a better idea for that keyword in question.

**Reply by HaloFour** [reply]
IMO `var` would be appropriate here, and the inferred type would be the natural lambda type, which would be implicitly convertible to any delegate type with a compatible signature.  The behavior of the compiler would effectively be the same as if the inferred type was an unspeakable synthetic delegate except that the compiler could optimize around not actually emitting the synthetic type or converting between delegate types.

**Reply by Rekkonnect** [reply]
That's a view I hadn't considered, and I'm fully supportive. This essentially enables usage of `var` for lambda expression assignments in all contexts, to which I say "why not already done?"

**Reply by 333fred** [reply]
var for lambdas is about to have a real meaning in C# 10. We almost certainly won't accept a different interpretation in a future version.

**Reply by HaloFour** [reply]
Indeed, I think the natural type feature for lambda will preclude this proposal.  I'm hopeful that future features that may enable signature equivalence in delegates combined with natural lambda types will effectively provide the same benefits, though.

**Comment by RikkiGibson** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/1694#discussioncomment-1352246
> Indeed, I think the natural type feature for lambda will preclude this proposal. I'm hopeful that future features that may enable signature equivalence in delegates combined with natural lambda types will effectively provide the same benefits, though.
